core/src/main/scala/cats/syntax/monadError.scala

```diff
@@ -32,6 +32,9 @@ final class MonadErrorOps[F[_], E, A](private val fa: F[A]) extends AnyVal {
 
   def redeemWith[B](recover: E => F[B], bind: A => F[B])(implicit F: MonadError[F, E]): F[B] =
     F.redeemWith[A, B](fa)(recover, bind)
+
+  def attemptTap[B](f: Either[E, A] => F[B])(implicit F: MonadError[F, E]): F[A] =
+    F.rethrow(F.flatTap(F.attempt(fa))(f))
```

Comment:
I'm not sure exactly what common pattern this is replacing? I guess exactly what the implementation is doing?

Reply:
We already have an analogue for `Future#onSuccess`/`Future#foreach` (`Monad#flatTap`) and `Future#onError` (`ApplicativeError#onError`). This would be the pure analogue to `Future#onComplete`, which is useful for use cases like reporting metrics or logging. This function makes call chains a bit more compact.

```scala
fa
  .attempt
  .flatTap {
    case Right(a) => log.info(s"succeeded with $a")
    case Left(e) => log.error(s"failed with $e")
  }
  .rethrow

// Ironically this one has fewer LOC but harder to factor out the bodies
fa
  .flatTap(a => log.info(s"succeeded with $a"))
  .onError(e => log.error(s"failed with $e"))

fa.attemptTap {
  case Right(a) => log.info(s"succeeded with $a")
  case Left(e) => log.error(s"failed with $e")
}
```

aside: I could see an argument for accepting `f` as a `PartialFunction`